explainer-datamodel.md

```diff
@@ -149,36 +159,207 @@ FormattedText.format( "The quick brown fox jumps over the lazy dog.", null, { wi
+runs. The iterator returned by `lines` has a read/write property called `inlineSize` which specifies where the
+next line's inlineSize break opportunity should be. It's default value is the value provided in the `lines` 
+function (parameter 3). The `inlineSize` property can be set to change the value used by the iterator. If the
+`inlineSize` value is changed, the change is "sticky" and will persist for subsequent lines until changed again.
+
+For example, if there is some external shape that text needs to wrap around, the lines iterator can be adjusted 
+at the appropriate times to shorten or lengthen the `inlineSize` constraint so that the lines can be positioned
+properly (line positioning must be done manually).
+
+![Image of a cat in the upper-left corner of a text box, with text flowing to the image's right side and continuing below it.](explainerresources/Available-Width.png)
+
+```js
+// Caller provides text to format, a CSS font string, width/height constraints,
+//  a box location/dimensions, and a rendering function callback 
+//  that accepts a `FormattedTextLine` (result from the lines iterator).
+function wrapAroundFloatLeftBox( text, cssFont, constraints, box = { width: 200, height: 200, marginRightBottom: 10 }, renderFunc ) {
+
+  // Prepare the input text and get a line iterator
+  //  (Set the default inlineSize for line-wrapping to the known constrained space.)
+  let formattedTextIterator = FormattedText.lines( text, cssFont, constraints.width - box.marginRightBottom + box.width );
+  
+  // sum of line metrics (collected from line instances)
+  let lineHeightTotal = 0;
+  // horizontal position offset for line placement (starts offset)
+  let offsetX = box.marginRightBottom + box.width;
+  
+  // Start iterating lines...
+  let { done, value: line } = formattedTextIterator.next();
+  while ( !done ) {
+    
+    // Render the line at the current offsets
+    renderFunc( line, offsetX, lineHeightTotal );
+    
+    // Prepare for the next line (using line metrics)
+    lineHeightTotal += line.height;
+        
+    // Determine if the offsetX position needs to change
+    if ( lineHeightTotal > ( box.marginRightBottom + box.height ) ) {
+      offsetX = 0;
+      formattedTextIterator.inlineSize = constraints.width; // Change the iterator's line-break position for subsequent lines...
+    }   
+   
+    // Request the next line (if available)
+    ( { done, value: line } = formattedTextIterator.next() );
+  }
+}
+```
+
+### Fast-forwarding of line iteration
+
+In some cases, the only tweaking that needs to be done comes early in an algorithm (such as in the 
+above algorithm, where once the line length (inlineSize) adjustment has been made to the iterator,
+the rest of the lines will not need any adjustments.
+
+The iterator can be quickly "fast-forwarded" by supplying the number of desired next lines in the 
+first parameter to the call to `next()`:
```

Comment:
I agree, but I think having a `next()` API is required in order to support the [JS iterator protocol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterator_protocol). Otherwise the `for..of` pattern won't work AFAIU.

Reply:
I was mostly referring to the _fast-forward_ behavior... is that a requirement of the iterator protocol?

Reply:
no, the iterator protocol just requires that the call to next (with no params) returns the next item in the iterator. I'm not sure if the iterator protocol even allows rewinding/fast forwarding, so we could be abusing it already, in which case perhaps the iterator protocol isn't a great fit--hoping additional feedback will reveal this.

_fast forward_ is just a convenience (I use in the multi-column balancing code example). It can be replaced with a simple for loop to collect the remaining lines...

Reply:
In offline conversation, we agreed to drop the fast-forward feature.

Reply:
Dropped in https://github.com/WICG/canvas-formatted-text/commit/1b63a0328276ef4cfb70e5e5f42711661436b0d6